### PR TITLE
Reviewer Alex - Astaire integration changes

### PR DESCRIPTION
**Pull request by bossmc**
Alex,

This is a combined pull request of various `cpp-common` changes related to Astaire integration.  Most of this has already been reviewed (in separate PRs) but can you please review the following sections of new code?

* `pdlog.h`
* `utils.h` and `utils.cpp`
* the marked change in `memcachedstore.cpp`

Thanks,

A

### CODE REVIEW

File: src/memcachedstore.cpp [hunk 1]
```diff
@@ -321,46 +245,14 @@ const std::vector<memcached_st*>& MemcachedStore::get_replicas(int vbucket,
...
-
-      LOG_DEBUG("Setting server to IP address %s port %s",
-                contact_details[0].c_str(),
-                contact_details[1].c_str());
-      int port = atoi(contact_details[1].c_str());
-      memcached_server_add(conn->st[_servers[ii]], contact_details[0].c_str(), port);
+      LOG_DEBUG("Setting server to IP address %s port %d",
+                server.c_str(),
+                port);
+      memcached_server_add(conn->st[_servers[ii]], server.c_str(), port);
```
Comment (bossmc) [inline src/memcachedstore.cpp]: This change (to use the new, common `split_host_port` function).

File: src/utils.cpp [hunk 2]
```diff
@@ -270,3 +270,55 @@ void Utils::hashToHex(unsigned char *hash_char, unsigned char *hex_char)
 // LCOV_EXCL_STOP
 
 bool Utils::StopWatch::_already_logged = false;
+
+bool Utils::split_host_port(const std::string& host_port,
+                            std::string& host,
+                            int& port)
+{
+  // The address is specified as either <IPv4 address>:<port>
+  // or [<IPv6 address>]:<port>.  Look for square brackets to determine whether
```
Comment (AlexHockey) [inline src/utils.cpp]: Or `localhost:port`?

File: src/utils.cpp [hunk 3]
```diff
@@ -270,3 +270,55 @@ void Utils::hashToHex(unsigned char *hash_char, unsigned char *hex_char)
...
+  std::vector<std::string> host_port_parts;
+  size_t close_bracket = host_port.find(']');
+
+  if (close_bracket == host_port.npos)
+  {
+    // IPv4 connection.  Split the string on the colon.
+    Utils::split_string(host_port, ':', host_port_parts);
+    if (host_port_parts.size() != 2)
+    {
+      LOG_ERROR("Malformed host/port %s", host_port.c_str());
```
Comment (AlexHockey) [inline src/utils.cpp]: Potential log spam.

File: src/utils.cpp [hunk 4]
```diff
@@ -270,3 +270,55 @@ void Utils::hashToHex(unsigned char *hash_char, unsigned char *hex_char)
...
+      return false;
+    }
+  }
+  else
+  {
+    // IPv6 connection.  Split the string on ']', which removes any white
+    // space from the start and the end, then remove the '[' from the
+    // start of the IP addreess string and the start of the ';' from the start
```
Comment (AlexHockey) [inline src/utils.cpp]: The start of the ';'??

File: src/utils.cpp [hunk 5]
```diff
@@ -270,3 +270,55 @@ void Utils::hashToHex(unsigned char *hash_char, unsigned char *hex_char)
...
+    // of the port string.
+    Utils::split_string(host_port, ']', host_port_parts);
+    if ((host_port_parts.size() != 2) ||
+        (host_port_parts[0][0] != '[') ||
+        (host_port_parts[1][0] != ':'))
+    {
+      LOG_ERROR("Malformed host/port %s", host_port.c_str());
```
Comment (AlexHockey) [inline src/utils.cpp]: Ditto

File: src/utils.cpp [hunk 6]
```diff
@@ -270,3 +270,55 @@ void Utils::hashToHex(unsigned char *hash_char, unsigned char *hex_char)
...
+    host_port_parts[1].erase(host_port_parts[1].begin());
+  }
+
+  port = atoi(host_port_parts[1].c_str());
+  host = host_port_parts[0];
+
+  // Check the port was parsed correctly.
+  if (std::to_string(port) != host_port_parts[1])
+  {
+    LOG_ERROR("Malformed port %s", host_port_parts[1].c_str());
```
Comment (AlexHockey) [inline src/utils.cpp]: Ditto

File: src/memcachedstore.cpp (hunk 1 above)
Comment (AlexHockey) [inline src/memcachedstore.cpp]: You old code skipped the server if the host/port was bad. I think you should check the return code from `split_host_port` and do the same thing.

**Comment by AlexHockey** [comment]
Reviewed.

File: src/memcachedstore.cpp (hunk 1 above)
Comment (bossmc) [inline src/memcachedstore.cpp]: Ah, indeed.

**Pull request merged**